src/odoc.ml

```diff
@@ -327,24 +327,53 @@ module Gen (S : sig val sctx : SC.t end) = struct
           (Path.Set.of_list (List.rev_append static_html html_files));
       end
 
-  let setup_pkg_html_rules =
-    let loaded = Package.Name.Table.create ~default_value:false in
-    fun ~pkg ~libs ->
-      if not (Package.Name.Table.get loaded pkg) then begin
-        Package.Name.Table.set loaded ~key:pkg ~data:true;
-        let requires = Lib.closure libs ~linking:false in
-        List.iter libs ~f:(setup_lib_html_rules ~requires);
-        let pkg_odocs = odocs (Pkg pkg) in
-        List.iter pkg_odocs ~f:(setup_html ~requires);
-        let odocs =
-          List.concat (
-            pkg_odocs
-            :: (List.map libs ~f:(fun lib -> odocs (Lib lib)))
-          ) in
-        let html_files = List.map ~f:(fun o -> o.html_file) odocs in
-        Build_system.Alias.add_deps (Dep.html_alias (Pkg pkg))
-          (Path.Set.of_list (List.rev_append static_html html_files))
-      end
+  let setup_pkg_html_rules_def =
+    let module Input = struct
+      type t = Package.Name.t * Lib.t list
+
+      let equal (p1, l1) (p2, l2) =
+        Package.Name.equal p1 p2
+        && List.equal Lib.equal l1 l2
+
+      let hash = Hashtbl.hash
```

Comment:
Using a custom `equal` with a polymorphic `Hashtbl.hash` is in general a bad idea.
This is broken exactly when polymorphic `(=)` is broken, and it's broken more badly (a hashtable will give inconsistent answers on whether or not an item is in there).
I think we should either define hash by hand or use the polymorphic equality. (much preferably the former)

Reply:
I agree it's broken, how should I define the hash by hand?

Reply:
Something like this should work:

```
let hash (p, ls) = Hashtbl.hash (Package.Name.hash p, List.hash Lib.hash ls)
```

Although, thinking more carefully about this example, I suspect it's not currently broken because there is never two physically-different Lib.t which are Lib.equal.

Reply:
For now, the following should do: `Hashtbl.hash (p, List.map ls ~f:Lib.id)`

Reply:
Thanks. How do I define `List.hash` then? like this?:

```
let hash f xs = Hashtbl.hash (map ~f xs)
```

Reply:
@diml I did something similar